### PR TITLE
add internal sharding

### PR DESCRIPTION
**Pull request by GusCaplan**
todo
- [x] fix ready events
- [x] fix up voice
- [x] fix up sending presence updates
- [x] **TEEEEST**
- this was tested on:
  - a user account with 100 guilds
  - a user account with 1 guild
  - a bot with 8 guilds
  - a bot with 16k guilds

### CODE REVIEW

**Comment by itslukej** [comment]
gus this is god-like, marry me

File: src/client/websocket/WebSocketShardManager.js [hunk 1]
```diff
@@ -35,14 +35,16 @@ class WebSocketShardManager extends EventEmitter {
...
     this.client.emit('debug', `Spawning ${this.shardCount} shard(s)`);
-    this.spawn(0);
-    const interval = setInterval(() => {
-      if (this.managers.size >= this.shardCount) {
-        clearInterval(interval);
-        return;
-      }
-      this.spawn(this.managers.size);
-    }, 5500);
+    const spawnLoop = (id) => {
```
Comment (hydrabolt) [inline src/client/websocket/WebSocketShardManager.js]: should just be `id => {`

File: src/client/websocket/WebSocketShardManager.js [hunk 2]
```diff
@@ -35,14 +35,16 @@ class WebSocketShardManager extends EventEmitter {
...
+      const manager = this.spawn(id);
+      manager.on('ready', () => {
```
Comment (hydrabolt) [inline src/client/websocket/WebSocketShardManager.js]: should be `.once(`, also try and have a fallback for failed connections to just continue with the next shard

File: src/client/Client.js [hunk 3]
```diff
@@ -24,9 +24,12 @@ class Client extends EventEmitter {
     super();
 
     // Obtain shard details from environment
-    if (!options.shardId && 'SHARD_ID' in process.env) options.shardId = Number(process.env.SHARD_ID);
+    if (!options.shardID && 'SHARD_ID' in process.env) options.shardID = Number(process.env.SHARD_ID);
```
Comment (Gawdl3y) [inline src/client/Client.js]: Breaking change, which is a no-go right now. If you add a fallback to the old `shardId` in addition to this, it's fine.

File: src/client/Client.js [hunk 4]
```diff
@@ -114,11 +117,17 @@ class Client extends EventEmitter {
      */
     this.presences = new Collection();
 
+    /**
+     * The authorization token for the logged in user/bot.
+     * @type {?string}
```
Comment (Gawdl3y) [inline src/client/Client.js]: Gotta add a `@name`

File: src/client/Client.js [hunk 5]
```diff
@@ -136,28 +145,18 @@ class Client extends EventEmitter {
...
 
   /**
    * The status for the logged in Client.
-   * @type {?number}
+   * @type {?Array<number>}
    * @readonly
    */
-  get status() {
-    return this.ws.status;
+  get statuses() {
```
Comment (Gawdl3y) [inline src/client/Client.js]: Breaking change.

File: src/client/Client.js [hunk 6]
```diff
@@ -170,12 +169,20 @@ class Client extends EventEmitter {
...
+  }
+
+  /**
+   * The average heartbeat ping of the websockets
    * @type {number}
    * @readonly
    */
   get ping() {
-    return this.pings.reduce((prev, p) => prev + p, 0) / this.pings.length;
+    return this.pings.reduce((a, b) => a + b, 0) / this.pings.length;
```
Comment (Gawdl3y) [inline src/client/Client.js]: Why? You managed to make the variable names *shittier*. 😒

File: src/client/Client.js [hunk 7]
```diff
@@ -257,12 +264,14 @@ class Client extends EventEmitter {
...
    */
-  syncGuilds(guilds = this.guilds) {
+  syncGuilds(guilds = this.guilds, shardID) {
     if (this.user.bot) return;
-    this.ws.send({
+    guilds = guilds.filter(g => g.shardID === shardID);
+    this.ws.managers.get(shardID).send({
       op: 12,
-      d: guilds instanceof Collection ? guilds.keyArray() : guilds.map(g => g.id),
+      d: guilds,
```
Comment (Gawdl3y) [inline src/client/Client.js]: ???
Sending a collection?!

File: src/client/ClientManager.js [hunk 8]
```diff
@@ -28,10 +22,18 @@ class ClientManager {
...
-    this.client.rest.methods.getGateway().then(gateway => {
+    this.client.rest.methods.getGateway(this.client.options.shardCount === 'auto').then(res => {
+      const gateway = `${res.url}/?v=${Constants.PROTOCOL_VERSION}`;
       this.client.emit(Constants.Events.DEBUG, `Using gateway ${gateway}`);
-      this.client.ws.connect(gateway);
+      if (res.shards) {
+        this.client.options.shardCount = res.shards;
+        this.client.emit(Constants.Events.DEBUG, `Using recommended shard count of ${res.shards}`);
+      }
+      const timeout = this.client.setTimeout(() =>
```
Comment (Gawdl3y) [inline src/client/ClientManager.js]: The `() =>` should be on the same line as the `reject()` call.

File: src/client/websocket/WebSocketManager.js [hunk 9]
```diff
@@ -130,16 +134,26 @@ class WebSocketManager extends EventEmitter {
...
   }
 
+  pong(startTime) {
+    this.pings.unshift(Date.now() - startTime);
+    if (this.pings.length > 3) this.pings.length = 3;
+    this.lastHeartbeatAck = true;
+  }
+
+  get ping() {
+    return (this.pings.reduce((prev, p) => prev + p, 0) / this.pings.length) || 0;
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketManager.js]: `|| 0` is stupid.

File: src/client/websocket/WebSocketManager.js [hunk 10]
```diff
@@ -333,13 +358,13 @@ class WebSocketManager extends EventEmitter {
   checkIfReady() {
     if (this.status !== Constants.Status.READY && this.status !== Constants.Status.NEARLY) {
       let unavailableCount = 0;
-      for (const guildID of this.client.guilds.keys()) {
+      for (const guildID of this.client.guilds.filter(g => g.shardID === this.shardID).keys()) {
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketManager.js]: Don't use `map()` here. Loop across the whole thing and test.

File: src/client/websocket/WebSocketManager.js [hunk 11]
```diff
@@ -333,13 +358,13 @@ class WebSocketManager extends EventEmitter {
...
         unavailableCount += this.client.guilds.get(guildID).available ? 0 : 1;
       }
       if (unavailableCount === 0) {
         this.status = Constants.Status.NEARLY;
         if (this.client.options.fetchAllMembers) {
-          const promises = this.client.guilds.map(g => g.fetchMembers());
+          const promises = this.client.guilds.filter(g => g.shardID === this.shardId).map(g => g.fetchMembers());
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketManager.js]: Ditto.

File: src/client/websocket/WebSocketShardManager.js [hunk 12]
```diff
@@ -0,0 +1,124 @@
+const EventEmitter = require('events').EventEmitter;
+const WebSocketManager = require('./WebSocketManager');
+const Constants = require('../../util/Constants');
+const PacketManager = require('./packets/WebSocketPacketManager');
+const Collection = require('../../util/Collection');
+
+class WebSocketShardManager extends EventEmitter {
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketShardManager.js]: Entire thing needs to be documented. All new code needs to be fully documented.

File: src/util/Constants.js [hunk 13]
```diff
@@ -83,7 +83,7 @@ const Endpoints = exports.Endpoints = {
   // general
   login: `${API}/auth/login`,
   logout: `${API}/auth/logout`,
-  gateway: `${API}/gateway`,
+  gateway: (bot) => `${API}/gateway${bot ? '/bot' : ''}`,
```
Comment (Gawdl3y) [inline src/util/Constants.js]: Breaking change, since constants are exposed.

**Comment by GusCaplan** [comment]
closing in favor of another thing because so many internal changes have been made in 11.1

**Pull request closed without merging**